### PR TITLE
Improve scroll behavior for active cells.

### LINKED ISSUE #478: focusing a cell by clicking scrolls to the bottom of the cell
If you make the output of the cell longer than a page, then this becomes frustrating, since the input scrolls away, even if you are editing the cell. This happens in chrome and firefox for me on OS X.

focus

### PR DESCRIPTION
**Pull request by jasongrout**
In this commit, we decouple setting a cell as active and scrolling to show it. This is useful when, for example, you are making a cell active by clicking on it. Clicking indicates a point of interest in the cell (like a position in the input editor, or a particular output slider you want to interact with), and scrolling when these click events make a cell active disrupts the user’s focus on the part they want to interact with. On the other hand, programmatically setting the active cell often does involve scrolling, so we added a method to explicitly scroll to the active cell.

Fixes #478. Also fixes the same long-standing problem with the current notebook.

### CODE REVIEW

**Comment by jasongrout** [comment]
CC @sccolbert

File: src/notebook/notebook/widget.ts [hunk 1]
```diff
@@ -739,32 +743,29 @@ class Notebook extends StaticNotebook {
...
       this.addClass(EDIT_CLASS);
       this.removeClass(COMMAND_CLASS);
-      if (widget) {
-        widget.focus();
-        if (widget instanceof MarkdownCellWidget) {
-          (widget as MarkdownCellWidget).rendered = false;
+      if (activeCell) {
+        activeCell.focus();
+        if (activeCell instanceof MarkdownCellWidget) {
+          (activeCell as MarkdownCellWidget).rendered = false;
```
Comment (sccolbert) [inline src/notebook/notebook/widget.ts]: this cast shouldn't be needed. `instanceof` is a type guard, so `activeCell` should be properly typed within the `if` block.

**Comment by sccolbert** [comment]
:+1:  LGTM modulo my comment

File: src/notebook/notebook/widget.ts (hunk 1 above)
Comment (jasongrout) [inline src/notebook/notebook/widget.ts]: Good catch!

File: src/utils.ts [hunk 2]
```diff
@@ -30,3 +30,82 @@ function findElement(parent: HTMLElement, className: string): HTMLElement {
...
+    }
+    if (er.top < ar.top && er.height > ar.height) {
+      area.scrollTop -= ar.bottom - er.bottom;
+      return;
+    }
+    if (er.bottom > ar.bottom && er.height < ar.height) {
+      area.scrollTop -= ar.bottom - er.bottom;
+      return;
+    }
+  }
```
Comment (sccolbert) [inline src/utils.ts]: do we typically have newline at EOF in JLab? If so, I'd flag this as missing one.

Comment (jasongrout) [inline src/utils.ts]: Yeah, I just noticed that too.

**Pull request merged**